### PR TITLE
Refactor and rearrange things.

### PR DESCRIPTION
**Pull request by chris-wood**

### CODE REVIEW

File: draft-irtf-cfrg-opaque.md [hunk 1]
```diff
@@ -1056,225 +1056,60 @@ Steps:
...
+The generic outline of OPAQUE with a 3-message AKE protocol includes three messages
+KE1, KE2, and KE3, where KE1 and KE2 include key exchange shares, e.g., DH values, sent
+by the client and server, respectively, and KE3 provides explicit client authentication and
+full forward security (without it, forward secrecy is only achieved against eavesdroppers
+which is insufficient for OPAQUE security). These messages are computed and exchanged with
+the following application APIs:
 
-The `ke1`, `ke2`, and `ke3` output values are the protocol messages sent between client and server.
+- ke1, client_state = ClientInit(client_identity, password, client_info)
+- ke2, client_info = ServerInit(server_identity, server_private_key, server_public_key, record, credential_identifier, oprf_seed, ke1)
```
Comment (bytemare) [inline draft-irtf-cfrg-opaque.md:1076]: No server state?

File: draft-irtf-cfrg-opaque.md [hunk 2]
```diff
@@ -1056,225 +1056,60 @@ Steps:
...
+Outputs `client_info`, `server_info` correspond to the optional information exchanged between
+client and server during the key exchange protocol. And finally, `session_key` and `export_key`
+are outputs to be consumed by applications. Applications can use `session_key` to derive
+additional keying material as needed.
 
-The following structure can be used by the client to carry internal values in a state between calls to `ClientInit()` and `ClientFinish()`:
+Both ClientFinish and ServerFinish return an error if authentication failed. In this case,
+clients and servers MUST NOT use any outputs from the protocol, such as `session_key` or
+`export_key`. ClientInit and ServerInit both implicitly return internal state objects
+`client_state` and `server_state`, respectively, with the following named fields:
```
Comment (bytemare) [inline draft-irtf-cfrg-opaque.md:1089]: If it's implicit, then maybe not return it at line 1075 :) (or add it at line 1076)

File: draft-irtf-cfrg-opaque.md [hunk 3]
```diff
@@ -1338,70 +1173,182 @@ client_mac
...
+                     I2OSP(len(client_identity), 2), client_identity,
+                     KE1,
+                     I2OSP(len(server_identity), 2), server_identity,
+                     KE2.inner_ke2)
+2. Output preamble
+~~~
+
+#### Shared Secret Derivation
+
+The OPAQUE-3DH shared secret derive during the key exchange protocol is computed
```
Comment (bytemare) [inline draft-irtf-cfrg-opaque.md:1228]:
```suggestion
The OPAQUE-3DH shared secret derived during the key exchange protocol is computed
```

File: draft-irtf-cfrg-opaque.md [hunk 4]
```diff
@@ -1338,70 +1173,182 @@ client_mac
...
+
+Using this shared secret, further keys used for encryption and authentication are
+computed using the following function.
+
+~~~
+DeriveKeys(ikm, preamble)
 
 Parameters:
-- Nn, the nonce length
+- Nx, the output size of the Extract function
```
Comment (bytemare) [inline draft-irtf-cfrg-opaque.md:1255]: This function doesn't use Nx directly. Is it necessary specifying it here?

File: draft-irtf-cfrg-opaque.md [hunk 5]
```diff
@@ -1338,70 +1173,182 @@ client_mac
...
 Input:
-- credential_request, a CredentialRequest structure
-- client_info, the optional client_info to send to the server. Note that it will not be encrypted, and only authenticated
+- client_identity, the optional encoded client identity, which is nil if not specified
+- password, an opaque byte string containing the client's password
+- client_info, optional application-specific information to exchange during the protocol
 
 Output:
-- ke1, a KE1 structure
+- KE1, a KE1 message structure
```
Comment (bytemare) [inline draft-irtf-cfrg-opaque.md:1291]:
```suggestion
- ke1, a KE1 message structure
```

File: draft-irtf-cfrg-opaque.md [hunk 6]
```diff
@@ -1056,225 +1056,60 @@ Steps:
...
-: The client's 3DH ephemeral secret key share.
-
-The following structure can be used by the server to carry internal values in a state between calls to `ServerInit()` and `ServerFinish()`:
-
-~~~
 struct {
     uint8 expected_client_mac[Nm];
     uint8 session_key[Nx];
+    KE1 ke1;
+    KE2 ke2;
```
Comment (bytemare) [inline draft-irtf-cfrg-opaque.md:1102]: why these two?

**Review by bytemare — CHANGES_REQUESTED** [review]

File: draft-irtf-cfrg-opaque.md (hunk 6 above)
Comment (chris-wood) [reply draft-irtf-cfrg-opaque.md:1102]: Oh, right, it doesn't need them!

File: draft-irtf-cfrg-opaque.md (hunk 4 above)
Comment (chris-wood) [reply draft-irtf-cfrg-opaque.md:1255]: Nope -- removed.

**Pull request merged**